gemstone/common/transform.py

```diff
@@ -50,3 +50,48 @@ def or_reduction(gen: Generator, sub_circuit_name: str, signal_name: str,
     gen.wire(read_data_reduce_or.ports.O, pass_through)
 
     return gen.ports[input_name]
+
+
+def replace(parent: Generator, old_gen: Generator, new_gen: Generator):
+    """
+    replace the old_gen with the new generator. The interfaces of
+    @old_gen has to be the same as the @new_gen
+
+    :param parent: parent generator that holds the old_gen
+    :param old_gen: target generator to be replaced
+    :param new_gen: new generator to use
+    :return: None
+    """
+    # we first make sure that the interfaces are the same
+    assert len(old_gen.ports) == len(new_gen.ports)
```

Comment:
we can simplify this to
```python
assert old_gen.ports.keys() == new_gen.ports.keys()
types = zip(*[[p.type() for p in gen.items()] for gen in (new_gen, old_gen)])
assert all([t[0] == t[1] for t in types])```

Reply:
As a not-python-power-user, this is very difficult for me to comprehend what's going on with the new code. I would prefer to keep it simple, stupid. On a philosophical sense,
> Everyone knows that debugging is twice as hard as writing a program in the first place. So if you're as clever as you can be when you write it, how will you ever debug it?
-- Brian Kernighan

Reply:
:D fair point. I'm ok with the verbose version